### PR TITLE
Add Read/Write::can_read/write_vectored

### PR DESCRIPTION
**Pull request by sfackler**
When working with an arbitrary reader or writer, code that uses vectored
operations may end up being slower than code that copies into a single
buffer when the underlying reader or writer doesn't actually support
vectored operations. These new methods allow you to ask the reader or
witer up front if vectored operations are efficiently supported.

Currently, you have to use some heuristics to guess by e.g. checking if
the read or write only accessed the first buffer. Hyper is one concrete
example of a library that has to do this dynamically:
https://github.com/hyperium/hyper/blob/0eaf304644a396895a4ce1f0146e596640bb666a/src/proto/h1/io.rs#L582-L594

### CODE REVIEW

**Comment by sfackler** [comment]
cc @nikomatsakis this is part of what we talked about when discussing buffer initialization and vectored IO.

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -580,6 +580,17 @@ pub trait Read {
         default_read_vectored(|b| self.read(b), bufs)
     }
 
+    /// Determines if this `Read`er has an efficient `read_vectored`
+    /// implementation.
+    ///
+    /// If a `Read`er does not override the default `read_vectored`
+    /// implementation, code using it may want to avoid the method all together
+    /// and coalesce writes into a single buffer for higher performance.
+    #[unstable(feature = "can_vector", issue = "none")]
```
Comment (sfackler) [inline src/libstd/io/mod.rs]: I'll fill in the issue assuming people are happy with this approach.

**Comment by nagisa** [comment]
Why not just transparently merge the buffers? At this point the support for vectored I/O is becoming complicated enough that I think a broader discussion is worthwhile (i.e. an RFC).

Also the name is not great. `can_*_vectored` implies that the feature will always fail to work at all if this returns `false` whereas the real intent of this function is just to signal native support and throughput capabilities of vectored I/O.

**Comment by sfackler** [comment]
> Why not just transparently merge the buffers?

Because user code has more flexibility around how to merge those buffers.

> Also the name is not great.

Sure - definitely open to better naming.

**Comment by KodrAus** [comment]
This seems reasonable to me.

> Also the name is not great.

Perhaps something along the lines of the runtime CPU checks?

```rust
let strategy = if writer.is_write_vectored_optimized() {
    WriteStrategy::Queue
} else {
    WriteStrategy::Flatten
};
```

Also not a really great name, but the `is_` prefix seems more in line with other branching methods.

**Comment by Dylan-DPC** [comment]
@sfackler any updates?

**Comment by sfackler** [comment]
`is_write_vectored_optimized` is a bit wordy, but works for me. It could be shortened to `is_write_vectored` potentially? It's all unstable so it doesn't matter too much right now.

**Comment by joelpalmer** [comment]
Triaged

**Comment by KodrAus** [comment]
I'd be on board with `is_*_vectored` too.

**Comment by sfackler** [comment]
Renamed, and added a tracking issue

**Comment by Dylan-DPC** [comment]
@sfackler you have a merge commit. Can you unmerge it and rebase instead? thanks

**Comment by sfackler** [comment]
Fixed

**Comment by Dylan-DPC** [comment]
@kodrAus this is ready for review

**Comment by Amanieu** [comment]
I feel that this is a rather ugly hack, but I can't really think of any better solution off the top of my head. I'm fine with trying it out on nightly for now.

r=me, but needs a rebase.
